### PR TITLE
feat: add previous hackathons field to application

### PR DESCRIPTION
**Pull request by Tony9984**
### Tickets:

-   HCK-

### List of changes:

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

- Add "previousHackathons" field to model, updated application validation, tests, route documentation, etc. to reflect changes

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] New release
-   [ ] This change requires a documentation update

### How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

-   [ ] Test A
-   [ ] Test B

**Test Configuration**:

**Firmware version:**
**Hardware:**
**Toolchain:**
**SDK:**

### Checklist:

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] My changes generate no new warnings
-   [ ] Listed change(s) in the Changelog
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [] New and existing unit tests pass locally with my changes
-   [x] I have made corresponding changes to the documentation
-   [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: middlewares/validators/validator.helper.js [hunk 1]
```diff
@@ -777,7 +791,7 @@ function searchSortValidator(fieldLocation, fieldName) {
             } else {
                 return false;
             }
-            if (!!model.searchableField(value)) {
+            if (model.searchableField(value)) {
```
Comment (logan-r) [inline middlewares/validators/validator.helper.js:794]: Was this a boolean cast? `!!` is odd in an if, should be casting already, just want to make sure we don't break anything here

File: package.json [hunk 2]
```diff
@@ -3,11 +3,11 @@
     "version": "3.0.0",
     "private": true,
     "scripts": {
```
Comment (logan-r) [inline package.json:5]: Can you exclude this file from the commit?

**Review by logan-r — COMMENTED** [review]
Looks good! Haven't run locally tho, can you confirm that:

a) Tests pass
b) You can POST `localhost:3000/api/hacker` with
```
{
    "accountId": "[AN EXISTING ACCOUNT ID]",
    "application":{
        "general":{
            "school": "McGill University",
            "degree": "Undergraduate",
            "fieldOfStudy": ["Computer Science"],
            "graduationYear": 2021,
            "jobInterest":"Internship",
            "URL":{
                "resume":"resumes/1543458163426-5bff4d736f86be0a41badb91",
                "github":"https://github.com/abcd",
                "dropler":"https://dribbble.com/abcd",
                "personal":"https://www.hi.com/",
                "linkedIn":"https://linkedin.com/in/abcd",
                "other":"https://github.com/hackmcgill/hackerAPI/issues/168"
            }
        },
        "shortAnswer": {
            "skills":["Javascript","Typescript"],
            "question1": "I love McHacks",
            "question2":"Pls accept me",
            "comments":"hi!",
            "previousHackathons": "0"
        },
        "other": {
            "gender": "male",
            "ethnicity": ["Asian or Pacific Islander"],
            "privacyPolicy": true,
            "codeOfConduct": true
        },
        "accommodation": {
            "travel": 0,
            "shirtSize": "S"
        }
    }
}
```
and it shows up in mongo

c) POSTing `localhost:3000/api/hacker` with
```
{
    "accountId": "[AN EXISTING ACCOUNT ID]",
    "application":{
        "general":{
            "school": "McGill University",
            "degree": "Undergraduate",
            "fieldOfStudy": ["Computer Science"],
            "graduationYear": 2021,
            "jobInterest":"Internship",
            "URL":{
                "resume":"resumes/1543458163426-5bff4d736f86be0a41badb91",
                "github":"https://github.com/abcd",
                "dropler":"https://dribbble.com/abcd",
                "personal":"https://www.hi.com/",
                "linkedIn":"https://linkedin.com/in/abcd",
                "other":"https://github.com/hackmcgill/hackerAPI/issues/168"
            }
        },
        "shortAnswer": {
            "skills":["Javascript","Typescript"],
            "question1": "I love McHacks",
            "question2":"Pls accept me",
            "comments":"hi!"
        },
        "other": {
            "gender": "male",
            "ethnicity": ["Asian or Pacific Islander"],
            "privacyPolicy": true,
            "codeOfConduct": true
        },
        "accommodation": {
            "travel": 0,
            "shirtSize": "S"
        }
    }
}
```
fails as it's missing previousHackathons

File: middlewares/validators/validator.helper.js (hunk 1 above)
Comment (Tony9984) [reply middlewares/validators/validator.helper.js:794]: not sure, i think the lint took it off by itself, should i restore it to !!?

File: package.json (hunk 2 above)
Comment (Tony9984) [reply package.json:5]: yeah sure! mybad haha

**Comment by Tony9984** [comment]
> Looks good! Haven't run locally tho, can you confirm that:
> 
> a) Tests pass
> b) You can POST `localhost:3000/api/hacker` with
> 
> ```
> {
>     "accountId": "[AN EXISTING ACCOUNT ID]",
>     "application":{
>         "general":{
>             "school": "McGill University",
>             "degree": "Undergraduate",
>             "fieldOfStudy": ["Computer Science"],
>             "graduationYear": 2021,
>             "jobInterest":"Internship",
>             "URL":{
>                 "resume":"resumes/1543458163426-5bff4d736f86be0a41badb91",
>                 "github":"https://github.com/abcd",
>                 "dropler":"https://dribbble.com/abcd",
>                 "personal":"https://www.hi.com/",
>                 "linkedIn":"https://linkedin.com/in/abcd",
>                 "other":"https://github.com/hackmcgill/hackerAPI/issues/168"
>             }
>         },
>         "shortAnswer": {
>             "skills":["Javascript","Typescript"],
>             "question1": "I love McHacks",
>             "question2":"Pls accept me",
>             "comments":"hi!",
>             "previousHackathons": "0"
>         },
>         "other": {
>             "gender": "male",
>             "ethnicity": ["Asian or Pacific Islander"],
>             "privacyPolicy": true,
>             "codeOfConduct": true
>         },
>         "accommodation": {
>             "travel": 0,
>             "shirtSize": "S"
>         }
>     }
> }
> ```
> 
> and it shows up in mongo
> 
> c) POSTing `localhost:3000/api/hacker` with
> 
> ```
> {
>     "accountId": "[AN EXISTING ACCOUNT ID]",
>     "application":{
>         "general":{
>             "school": "McGill University",
>             "degree": "Undergraduate",
>             "fieldOfStudy": ["Computer Science"],
>             "graduationYear": 2021,
>             "jobInterest":"Internship",
>             "URL":{
>                 "resume":"resumes/1543458163426-5bff4d736f86be0a41badb91",
>                 "github":"https://github.com/abcd",
>                 "dropler":"https://dribbble.com/abcd",
>                 "personal":"https://www.hi.com/",
>                 "linkedIn":"https://linkedin.com/in/abcd",
>                 "other":"https://github.com/hackmcgill/hackerAPI/issues/168"
>             }
>         },
>         "shortAnswer": {
>             "skills":["Javascript","Typescript"],
>             "question1": "I love McHacks",
>             "question2":"Pls accept me",
>             "comments":"hi!"
>         },
>         "other": {
>             "gender": "male",
>             "ethnicity": ["Asian or Pacific Islander"],
>             "privacyPolicy": true,
>             "codeOfConduct": true
>         },
>         "accommodation": {
>             "travel": 0,
>             "shirtSize": "S"
>         }
>     }
> }
> ```
> 
> fails as it's missing previousHackathons

got it! here's what i have, everything should be good:

a) all tests pass! except for the 2 that i sent you on slack
b) returns a 500, debugging it right now
c) works as intended! 422 error, with previousHackathons as missing field

**Comment by pierreTklein** [comment]
Are you able to debug the 500 error?

**Comment by logan-r** [comment]
Ran tests on latest `dev` code & tests pass, meaning contrary to my thinking earlier on slack, this is a bug introduced by this PR

@Tony9984 Finishing up a midterm today, can help debug tomorrow if this is still an issue

image

**Comment by Tony9984** [comment]
> Are you able to debug the 500 error?

**Comment by Tony9984** [comment]
hey guys! thanks for the updates! tried debugging it, couldn't find the issue, I'll try again tonight after my midterm. @logan-r good luck for yours!! hope everything goes well.

**Review by logan-r — APPROVED** [review]
Looks great!!

**Pull request merged**